src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,53 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private int size;
+    private Node<T> head;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+
+        Node(T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+        }
+
+        Node(T value) {
+            this.value = value;
+        }
+    }
 
+    public void push(T value) {
+        head = (Node<T>) new Node(value, head);
+        size++;
     }
 
     public T peek() {
-        return null;
+        checkForHead();
+        return head.value;
     }
 
     public T pop() {
-        return null;
+        checkForHead();
+        size--;
+        T deletedValue = head.value;
+        if (head.next != null) {
```

Comment:
it looks like we don't need this if, if it's null you will just set head to null but you check in your checkForHead anyway